components/gantry/oneaxis/setup.go

```diff
@@ -0,0 +1,390 @@
+	g.limitsPolled = true
+	g.limitHitMu.Unlock()
+
+	var errs error
+	for {
+		select {
+		case <-g.ctx.Done():
+			return
+		default:
+		}
+		for idx := range g.limitSwitchPins {
+			hit, err := g.limitHit(g.ctx, idx)
+			if hit || err != nil {
+				g.limitHitMu.Lock()
+				g.limitsPolled = false
+				g.limitHitMu.Unlock()
+				errs := multierr.Combine(errs, err)
+				// motor operation manager cancels running
+				err = g.oAx.motor.Stop(g.ctx, nil)
+				errs = multierr.Combine(errs, err)
+				if errs != nil {
+					g.logger.Error(errs)
+					return
+				}
+			}
+		}
+	}
+}
+
+func (g *limitSwitchGantry) limitHit(ctx context.Context, offset int) (bool, error) {
+	if offset < 0 || offset > 1 {
+		return true, errBadNumLimitSwitches(g.oAx.name, offset)
+	}
+	pin, err := g.oAx.board.GPIOPinByName(g.limitSwitchPins[offset])
+	if err != nil {
+		return false, err
+	}
+	high, err := pin.Get(ctx, nil)
+
+	return high == g.limitHigh, err
+}
+
+func (g *limitSwitchGantry) testLimit(ctx context.Context, limInt int) (float64, error) {
+	defer utils.UncheckedErrorFunc(func() error {
+		return g.oAx.motor.Stop(ctx, nil)
+	})
+
+	d := -1.0 // go backwards to hit first limit switch pin
+	if limInt > 0 {
+		d *= -1 // go forwards to hit second limit switch pin
+	}
+
+	err := g.oAx.motor.GoFor(ctx, d*g.oAx.rpm, 0, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	start := time.Now()
+	for {
+		hit, err := g.limitHit(ctx, limInt)
+		if err != nil {
+			return 0, err
+		}
+		if hit {
+			err = g.oAx.motor.Stop(ctx, nil)
+			if err != nil {
+				return 0, err
+			}
+			break
+		}
+
+		elapsed := start.Sub(start)
+		if elapsed > (time.Second * 15) {
+			return 0, errors.New("gantry timed out testing limit")
+		}
+
+		if !utils.SelectContextOrWait(ctx, time.Millisecond*10) {
+			return 0, ctx.Err()
+		}
+	}
+
+	return g.oAx.motor.Position(ctx, nil)
+}
+
+func (g *limitSwitchGantry) homeWithLimSwitch(ctx context.Context, limSwitchPins []string) error {
+	ctx, done := g.oAx.opMgr.New(ctx)
+	defer done()
+	positionA, err := g.testLimit(ctx, 0)
+	if err != nil {
+		return err
+	}
+
+	np := len(g.limitSwitchPins)
+	var positionB float64
+	switch len(limSwitchPins) {
+	case 1:
+		if g.oAx.lengthMm == 0.0 || g.oAx.mmPerRevolution == 0.0 {
+			return errDimensionsNotFound(g.oAx.name, g.oAx.lengthMm, g.oAx.mmPerRevolution)
+		}
+		revPerLength := g.oAx.lengthMm / g.oAx.mmPerRevolution
+		positionB = positionA + revPerLength
+	case 2:
+		positionB, err = g.testLimit(ctx, 1)
+		if err != nil {
+			return err
+		}
+	default:
+		return errBadNumLimitSwitches(g.oAx.name, np)
+	}
+
+	g.oAx.logger.Debugf("positionA: %0.2f positionB: %0.2f", positionA, positionB)
+
+	g.oAx.positionLimits = []float64{positionA, positionB}
+
+	if g.oAx.gantryRange = positionB - positionA; g.oAx.gantryRange == 0 {
+		return errZeroLengthGantry
+	}
+
+	// Go backwards so limit stops are not hit.
+	x := g.oAx.linearToRotational(0.9 * (g.oAx.gantryRange + positionA))
```

Comment:
If your gantry has 2 limit switches, this makes perfect sense: we're near the far end, so just back off 10%. but if it only has 1 limit switch, this is going to travel nearly the entire range from the near end. Mild preference to make this slightly more intelligent and go 10% from whichever end it's nearest.

```go
target := 0.9
if len(limSwitchPins) == 1 {
    target = 0.1  // We're near the start, not the end
}
x := g.oAx.linearToRotational(target * g.oAx.gantryRange + positionA)

Reply:
Oof, I missed this at first! Remove the innermost parentheses: the 0.9 should only multiply the range, not the starting position.

As an example of trouble, imagine that positionA is a million and gantryRange is a hundred.

Reply:
really nice catch, thank you.